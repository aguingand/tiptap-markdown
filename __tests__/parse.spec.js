import { describe, test, expect } from "vitest";
import { parse, dedent } from './utils';

describe('parse', () => {
    describe('marks', () => {
        describe('text', () => {
            test('text', () => {
                expect(parse('example')).toMatchSnapshot();
            });
            test('link', () => {
                expect(parse('http://example.org')).toMatchSnapshot();
            });
            test('soft break', () => {
                expect(parse('example1\nexample2')).toMatchSnapshot();
            });
        })
        describe('bold', () => {
            test('markdown', () => {
                expect(parse('**example**')).toMatchSnapshot();
            });
            test('html', () => {
                expect(parse('<b>example</b>')).toMatchSnapshot();
            });
        });
        describe('italic', () => {
            test('markdown', () => {
                expect(parse('*example*')).toMatchSnapshot();
            });
            test('html', () => {
                expect(parse('<em>example</em>')).toMatchSnapshot();
            });
        });
        describe('strike', () => {
            test('markdown', () => {
                expect(parse('~~example~~')).toMatchSnapshot();
            });
            test('html', () => {
                expect(parse('<s>example</s>')).toMatchSnapshot();
            });
        });
        describe('code', () => {
            test('markdown', () => {
                expect(parse('`example`')).toMatchSnapshot();
            });
            test('html', () => {
                expect(parse('<code>example</code>')).toMatchSnapshot();
            });
        });
        describe('link', () => {
            test('markdown', () => {
                expect(parse('[example](http://example.org)')).toMatchSnapshot();
            });
            test('markdown with linkify', () => {
                expect(parse('http://example.org', { linkify:true })).toMatchSnapshot();
            });
            test('html', () => {
                expect(parse('<a href="http://example.org">example</a>')).toMatchSnapshot();
            });
        });
    });
    describe('nodes', () => {
        describe('paragraph', () => {
            test('markdown', () => {
                expect(parse('example1\n\nexample2')).toMatchSnapshot();
            });
            test('html', () => {
                expect(parse('<p>example1</p><p>example2</p>')).toMatchSnapshot();
            });
        });
        describe('headings', () => {
            test('markdown h1', () => {
                expect(parse('# example')).toMatchSnapshot();
            });
            test('markdown h2', () => {
                expect(parse('## example')).toMatchSnapshot();
            });
            test('markdown h3', () => {
                expect(parse('### example')).toMatchSnapshot();
            });
            test('markdown h4', () => {
                expect(parse('#### example')).toMatchSnapshot();
            });
            test('markdown h5', () => {
                expect(parse('##### example')).toMatchSnapshot();
            });
            test('markdown h6', () => {
                expect(parse('###### example')).toMatchSnapshot();
            });
            test('html h1', () => {
                expect(parse('<h1>example</h1>')).toMatchSnapshot();
            });
        });
        describe('bullet list', () => {
            test('markdown marker `-`', () => {
                expect(parse('- example1\n\n- example2')).toMatchSnapshot();
            });
            test('markdown marker `*`', () => {
                expect(parse('* example1\n\n* example2')).toMatchSnapshot();
            });
            test('html', () => {
                expect(parse('<ul><li>example1</li><li>example2</li></ul>')).toMatchSnapshot();
            });
        });
        describe('ordered list', () => {
            test('markdown', () => {
                expect(parse('1. example1\n2. example2')).toMatchSnapshot();
            });
            test('html', () => {
                expect(parse('<ol><li>example1</li><li>example2</li></ol>')).toMatchSnapshot();
            });
        });
        describe('fence', () => {
            test('markdown', () => {
                expect(parse('```\nexample\n```')).toMatchSnapshot();
            });
            test('markdown with lang', () => {
                expect(parse('```js\nexample\n```')).toMatchSnapshot();
            });
            test('markdown with languageClassPrefix', () => {
                expect(parse('```js\nexample\n```', { codeBlock: { languageClassPrefix: 'lang--' } }, true))
                    .toEqual('<pre><code class="lang--js">example</code></pre>');
            })
        });
        describe('code block', () => {
            test('markdown', () => {
                expect(parse('    example')).toMatchSnapshot();
            });
            test('html', () => {
                expect(parse('<pre><code>example</code></pre>')).toMatchSnapshot();
            });
        });
        describe('image', () => {
            test('markdown', () => {
                expect(parse('![example](example.jpg)')).toMatchSnapshot();
            });
            test('markdown inline', () => {
                expect(parse('![example](example.jpg)', { image: { inline: true } })).toMatchSnapshot();
            });
            test('html', () => {
                expect(parse('<img src="example.jpg" alt="example">')).toMatchSnapshot();
            });
        });
        describe('hr', () => {
            test('markdown', () => {
                expect(parse('---')).toMatchSnapshot();
            })
            test('html', () => {
                expect(parse('<hr>')).toMatchSnapshot();
            });
        });
        describe('hard break', () => {
            test('markdown', () => {
                expect(parse('example1  \nexample2')).toMatchSnapshot();
            })
            test('markdown with breaks option', () => {
                expect(parse('example1\nexample2', { breaks: true })).toMatchSnapshot();
            })
            test('markdown with breaks option + inline', () => {
                expect(parse('example1\nexample2', { breaks: true, inline: true })).toMatchSnapshot();
            })
            test('html', () => {
                expect(parse('example1<br>example2')).toMatchSnapshot();
            });
        });
        describe('table', () => {
            test('markdown', () => {
                expect(parse(dedent`
                    example1 | example2
                    --- | ---
                    example3 | example4
                `)).toMatchSnapshot();
            });

            test('html', () => {
                expect(parse(dedent`
                    <table>
                    <thead>
                        <tr>
                            <th>example1</th>
                            <th>example2</th>
                        </tr>
                    </thead>
                    <tbody>
                        <tr>
                            <td>example3</td>
                            <td>example4</td>
                        </tr>
                    </tbody>
                    </table>
                `)).toMatchSnapshot();
            });
        });
        describe('html', () => {
            test('block', () => {
                expect(parse('<custom-element>example</custom-element>', {
                    htmlNode: {
                        group: 'block',
                        content: 'inline*',
                        parseHTML: () => [{
                            tag: 'custom-element',
                        }],
                    },
                })).toMatchSnapshot();
            });
            test('inline', () => {
                expect(parse('<custom-element></custom-element>', {
                    htmlNode: {
                        group: 'inline',
                        inline: true,
                        parseHTML: () => [{
                            tag: 'custom-element',
                        }],
                    },
                })).toMatchSnapshot();
            });
            test('disabled', () => {
                expect(parse('<custom-element></custom-element>', {
                    html: false,
                    htmlNode: {
                        group: 'block',
                        parseHTML: () => [{
                            tag: 'custom-element',
                        }],
                    },
                })).toMatchSnapshot();
            })
        });
    });
    describe('options', () => {
        describe('inline', () => {
            test('text', () => {
                expect(parse('example', { inline: true })).toMatchSnapshot();
            });
            test('text with spaces', () => {
                expect(parse('  example ', { inline: true })).toMatchSnapshot();
            });
        });
    });
});

