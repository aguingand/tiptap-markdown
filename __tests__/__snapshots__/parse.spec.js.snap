// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parse > marks > bold > html 1`] = `
[
  {
    "content": [
      {
        "marks": [
          {
            "type": "bold",
          },
        ],
        "text": "example",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse > marks > bold > markdown 1`] = `
[
  {
    "content": [
      {
        "marks": [
          {
            "type": "bold",
          },
        ],
        "text": "example",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse > marks > code > html 1`] = `
[
  {
    "content": [
      {
        "marks": [
          {
            "type": "code",
          },
        ],
        "text": "example",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse > marks > code > markdown 1`] = `
[
  {
    "content": [
      {
        "marks": [
          {
            "type": "code",
          },
        ],
        "text": "example",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse > marks > italic > html 1`] = `
[
  {
    "content": [
      {
        "marks": [
          {
            "type": "italic",
          },
        ],
        "text": "example",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse > marks > italic > markdown 1`] = `
[
  {
    "content": [
      {
        "marks": [
          {
            "type": "italic",
          },
        ],
        "text": "example",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse > marks > link > html 1`] = `
[
  {
    "content": [
      {
        "marks": [
          {
            "attrs": {
              "class": null,
              "href": "http://example.org",
              "target": "_blank",
            },
            "type": "link",
          },
        ],
        "text": "example",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse > marks > link > markdown 1`] = `
[
  {
    "content": [
      {
        "marks": [
          {
            "attrs": {
              "class": null,
              "href": "http://example.org",
              "target": "_blank",
            },
            "type": "link",
          },
        ],
        "text": "example",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse > marks > link > markdown with linkify 1`] = `
[
  {
    "content": [
      {
        "marks": [
          {
            "attrs": {
              "class": null,
              "href": "http://example.org",
              "target": "_blank",
            },
            "type": "link",
          },
        ],
        "text": "http://example.org",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse > marks > strike > html 1`] = `
[
  {
    "content": [
      {
        "marks": [
          {
            "type": "strike",
          },
        ],
        "text": "example",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse > marks > strike > markdown 1`] = `
[
  {
    "content": [
      {
        "marks": [
          {
            "type": "strike",
          },
        ],
        "text": "example",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse > marks > text > link 1`] = `
[
  {
    "content": [
      {
        "text": "http://example.org",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse > marks > text > soft break 1`] = `
[
  {
    "content": [
      {
        "text": "example1 example2",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse > marks > text > text 1`] = `
[
  {
    "content": [
      {
        "text": "example",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse > nodes > bullet list > html 1`] = `
[
  {
    "attrs": {
      "tight": true,
    },
    "content": [
      {
        "content": [
          {
            "content": [
              {
                "text": "example1",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
      {
        "content": [
          {
            "content": [
              {
                "text": "example2",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
    ],
    "type": "bulletList",
  },
]
`;

exports[`parse > nodes > bullet list > markdown marker \`*\` 1`] = `
[
  {
    "attrs": {
      "tight": false,
    },
    "content": [
      {
        "content": [
          {
            "content": [
              {
                "text": "example1",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
      {
        "content": [
          {
            "content": [
              {
                "text": "example2",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
    ],
    "type": "bulletList",
  },
]
`;

exports[`parse > nodes > bullet list > markdown marker \`-\` 1`] = `
[
  {
    "attrs": {
      "tight": false,
    },
    "content": [
      {
        "content": [
          {
            "content": [
              {
                "text": "example1",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
      {
        "content": [
          {
            "content": [
              {
                "text": "example2",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
    ],
    "type": "bulletList",
  },
]
`;

exports[`parse > nodes > code block > html 1`] = `
[
  {
    "attrs": {
      "language": null,
    },
    "content": [
      {
        "text": "example",
        "type": "text",
      },
    ],
    "type": "codeBlock",
  },
]
`;

exports[`parse > nodes > code block > markdown 1`] = `
[
  {
    "attrs": {
      "language": null,
    },
    "content": [
      {
        "text": "example",
        "type": "text",
      },
    ],
    "type": "codeBlock",
  },
]
`;

exports[`parse > nodes > fence > markdown 1`] = `
[
  {
    "attrs": {
      "language": null,
    },
    "content": [
      {
        "text": "example",
        "type": "text",
      },
    ],
    "type": "codeBlock",
  },
]
`;

exports[`parse > nodes > fence > markdown with lang 1`] = `
[
  {
    "attrs": {
      "language": "js",
    },
    "content": [
      {
        "text": "example",
        "type": "text",
      },
    ],
    "type": "codeBlock",
  },
]
`;

exports[`parse > nodes > hard break > html 1`] = `
[
  {
    "content": [
      {
        "text": "example1",
        "type": "text",
      },
      {
        "type": "hardBreak",
      },
      {
        "text": "example2",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse > nodes > hard break > markdown 1`] = `
[
  {
    "content": [
      {
        "text": "example1",
        "type": "text",
      },
      {
        "type": "hardBreak",
      },
      {
        "text": "example2",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse > nodes > hard break > markdown with breaks option + inline 1`] = `
[
  {
    "text": "example1",
    "type": "text",
  },
  {
    "type": "hardBreak",
  },
  {
    "text": "example2",
    "type": "text",
  },
]
`;

exports[`parse > nodes > hard break > markdown with breaks option 1`] = `
[
  {
    "content": [
      {
        "text": "example1",
        "type": "text",
      },
      {
        "type": "hardBreak",
      },
      {
        "text": "example2",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse > nodes > headings > html h1 1`] = `
[
  {
    "attrs": {
      "level": 1,
    },
    "content": [
      {
        "text": "example",
        "type": "text",
      },
    ],
    "type": "heading",
  },
]
`;

exports[`parse > nodes > headings > markdown h1 1`] = `
[
  {
    "attrs": {
      "level": 1,
    },
    "content": [
      {
        "text": "example",
        "type": "text",
      },
    ],
    "type": "heading",
  },
]
`;

exports[`parse > nodes > headings > markdown h2 1`] = `
[
  {
    "attrs": {
      "level": 2,
    },
    "content": [
      {
        "text": "example",
        "type": "text",
      },
    ],
    "type": "heading",
  },
]
`;

exports[`parse > nodes > headings > markdown h3 1`] = `
[
  {
    "attrs": {
      "level": 3,
    },
    "content": [
      {
        "text": "example",
        "type": "text",
      },
    ],
    "type": "heading",
  },
]
`;

exports[`parse > nodes > headings > markdown h4 1`] = `
[
  {
    "attrs": {
      "level": 4,
    },
    "content": [
      {
        "text": "example",
        "type": "text",
      },
    ],
    "type": "heading",
  },
]
`;

exports[`parse > nodes > headings > markdown h5 1`] = `
[
  {
    "attrs": {
      "level": 5,
    },
    "content": [
      {
        "text": "example",
        "type": "text",
      },
    ],
    "type": "heading",
  },
]
`;

exports[`parse > nodes > headings > markdown h6 1`] = `
[
  {
    "attrs": {
      "level": 6,
    },
    "content": [
      {
        "text": "example",
        "type": "text",
      },
    ],
    "type": "heading",
  },
]
`;

exports[`parse > nodes > hr > html 1`] = `
[
  {
    "type": "horizontalRule",
  },
]
`;

exports[`parse > nodes > hr > markdown 1`] = `
[
  {
    "type": "horizontalRule",
  },
]
`;

exports[`parse > nodes > html > block 1`] = `
[
  {
    "content": [
      {
        "text": "example",
        "type": "text",
      },
    ],
    "type": "html-node",
  },
]
`;

exports[`parse > nodes > html > disabled 1`] = `
[
  {
    "content": [
      {
        "text": <custom-element>
</custom-element>,
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse > nodes > html > inline 1`] = `
[
  {
    "content": [
      {
        "type": "html-node",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse > nodes > image > html 1`] = `
[
  {
    "attrs": {
      "alt": "example",
      "src": "example.jpg",
      "title": null,
    },
    "type": "image",
  },
]
`;

exports[`parse > nodes > image > markdown 1`] = `
[
  {
    "attrs": {
      "alt": "example",
      "src": "example.jpg",
      "title": null,
    },
    "type": "image",
  },
]
`;

exports[`parse > nodes > image > markdown inline 1`] = `
[
  {
    "content": [
      {
        "attrs": {
          "alt": "example",
          "src": "example.jpg",
          "title": null,
        },
        "type": "image",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse > nodes > ordered list > html 1`] = `
[
  {
    "attrs": {
      "start": 1,
      "tight": true,
    },
    "content": [
      {
        "content": [
          {
            "content": [
              {
                "text": "example1",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
      {
        "content": [
          {
            "content": [
              {
                "text": "example2",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
    ],
    "type": "orderedList",
  },
]
`;

exports[`parse > nodes > ordered list > markdown 1`] = `
[
  {
    "attrs": {
      "start": 1,
      "tight": true,
    },
    "content": [
      {
        "content": [
          {
            "content": [
              {
                "text": "example1",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
      {
        "content": [
          {
            "content": [
              {
                "text": "example2",
                "type": "text",
              },
            ],
            "type": "paragraph",
          },
        ],
        "type": "listItem",
      },
    ],
    "type": "orderedList",
  },
]
`;

exports[`parse > nodes > paragraph > html 1`] = `
[
  {
    "content": [
      {
        "text": "example1",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  {
    "content": [
      {
        "text": "example2",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse > nodes > paragraph > markdown 1`] = `
[
  {
    "content": [
      {
        "text": "example1",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
  {
    "content": [
      {
        "text": "example2",
        "type": "text",
      },
    ],
    "type": "paragraph",
  },
]
`;

exports[`parse > nodes > table > html 1`] = `
[
  {
    "content": [
      {
        "content": [
          {
            "attrs": {
              "colspan": 1,
              "colwidth": null,
              "rowspan": 1,
            },
            "content": [
              {
                "content": [
                  {
                    "text": "example1",
                    "type": "text",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "tableHeader",
          },
          {
            "attrs": {
              "colspan": 1,
              "colwidth": null,
              "rowspan": 1,
            },
            "content": [
              {
                "content": [
                  {
                    "text": "example2",
                    "type": "text",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "tableHeader",
          },
        ],
        "type": "tableRow",
      },
      {
        "content": [
          {
            "attrs": {
              "colspan": 1,
              "colwidth": null,
              "rowspan": 1,
            },
            "content": [
              {
                "content": [
                  {
                    "text": "example3",
                    "type": "text",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "tableCell",
          },
          {
            "attrs": {
              "colspan": 1,
              "colwidth": null,
              "rowspan": 1,
            },
            "content": [
              {
                "content": [
                  {
                    "text": "example4",
                    "type": "text",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "tableCell",
          },
        ],
        "type": "tableRow",
      },
    ],
    "type": "table",
  },
]
`;

exports[`parse > nodes > table > markdown 1`] = `
[
  {
    "content": [
      {
        "content": [
          {
            "attrs": {
              "colspan": 1,
              "colwidth": null,
              "rowspan": 1,
            },
            "content": [
              {
                "content": [
                  {
                    "text": "example1",
                    "type": "text",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "tableHeader",
          },
          {
            "attrs": {
              "colspan": 1,
              "colwidth": null,
              "rowspan": 1,
            },
            "content": [
              {
                "content": [
                  {
                    "text": "example2",
                    "type": "text",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "tableHeader",
          },
        ],
        "type": "tableRow",
      },
      {
        "content": [
          {
            "attrs": {
              "colspan": 1,
              "colwidth": null,
              "rowspan": 1,
            },
            "content": [
              {
                "content": [
                  {
                    "text": "example3",
                    "type": "text",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "tableCell",
          },
          {
            "attrs": {
              "colspan": 1,
              "colwidth": null,
              "rowspan": 1,
            },
            "content": [
              {
                "content": [
                  {
                    "text": "example4",
                    "type": "text",
                  },
                ],
                "type": "paragraph",
              },
            ],
            "type": "tableCell",
          },
        ],
        "type": "tableRow",
      },
    ],
    "type": "table",
  },
]
`;

exports[`parse > options > inline > text 1`] = `
[
  {
    "text": "example",
    "type": "text",
  },
]
`;

exports[`parse > options > inline > text with spaces 1`] = `
[
  {
    "text": "  example ",
    "type": "text",
  },
]
`;
